scene/gui/text_edit.cpp

```diff
@@ -2406,6 +2455,78 @@ bool TextEdit::is_text_field() const {
 	return true;
 }
 
+Variant TextEdit::get_drag_data(const Point2 &p_point) {
+	if (selection.active && selection.drag_attempt) {
+		String t = get_selected_text();
+		Label *l = memnew(Label);
+		l->set_text(t);
+		set_drag_preview(l);
+		return t;
+	}
+
+	return Variant();
+}
+
+bool TextEdit::can_drop_data(const Point2 &p_point, const Variant &p_data) const {
+	bool drop_override = Control::can_drop_data(p_point, p_data); // In case user wants to drop custom data.
+	if (drop_override) {
+		return drop_override;
+	}
+
+	return is_editable() && p_data.get_type() == Variant::STRING;
+}
+
+void TextEdit::drop_data(const Point2 &p_point, const Variant &p_data) {
+	Control::drop_data(p_point, p_data);
+
+	if (is_editable() && p_data.get_type() == Variant::STRING) {
+		Point2i pos = get_line_column_at_pos(get_local_mouse_pos());
+		int caret_row_tmp = pos.y;
+		int caret_column_tmp = pos.x;
+		if (selection.drag_attempt) {
+			selection.drag_attempt = false;
+			if (!is_mouse_over_selection(!Input::get_singleton()->is_key_pressed(Key::CTRL))) {
+				begin_complex_operation();
+				if (!Input::get_singleton()->is_key_pressed(Key::CTRL)) {
+					if (caret_row_tmp > selection.to_line) {
+						caret_row_tmp = caret_row_tmp - (selection.to_line - selection.from_line);
+					} else if (caret_row_tmp == selection.to_line && caret_column_tmp >= selection.to_column) {
+						caret_column_tmp = caret_column_tmp - (selection.to_column - selection.from_column);
+					}
+					delete_selection();
+				} else {
+					deselect();
+				}
+
+				set_caret_line(caret_row_tmp, true, false);
+				set_caret_column(caret_column_tmp);
+				insert_text_at_caret(p_data);
+				end_complex_operation();
```

Comment:
We shouldn't need to call `delete_selection` here as `insert_text_at_caret` will handle that. This should also remove the need to make this a complex operation. Similar for the `else if` below.

Reply:
I've removed the second `delete_selection` after the `else if`
The first one instead is needed, because is true that `insert_text_at_caret` call `delete_selection`, but `delete_selection` reset the caret position at the beginning of the selection and this break the code for the drag and drop.